shell/plugins/steve/worker/web-worker.advanced.js

```diff
@@ -131,21 +154,91 @@ const workerActions = {
 
       state.watcher.connect(connectionMetadata);
 
-      // Flush the workerQueue
-      while (state.workerQueue.length > 0) {
-        trace('createWatcher', 'flushing workerQueue', state.workerQueue);
+      // Flush the watcherQueue
+      workerActions.flushWatcherQueue();
+    }
+  },
+  waitingForResponse: (request) => {
+    if (!caches[SCHEMA]) {
+      state.apiQueue.push({ waitingForResponse: request });
 
-        const workerMessage = state.workerQueue.shift();
-        const [action, msg] = Object.entries(workerMessage)[0];
+      return;
+    }
+    const { params, requestHash } = request;
 
-        if (workerActions[action]) {
-          workerActions[action](msg);
-        } else {
-          console.warn('no associated action for:', action); // eslint-disable-line no-console
-        }
+    workerActions.request(params, response => self.postMessage({ awaitedResponse: { response, requestHash } }));
+  },
+  request: (params, resolver = () => {}) => {
+    const {
+      type, namespace, id, filter, sortBy, sortOrder, limit
+    } = params;
+
+    if (!caches[type]) {
+      state.api.request(params)
+        .then((res) => {
+          resolver(res);
+        });
+    } else {
+      caches[type].find({
```

Comment:
I've still got to test error handling, but this is one of the place where something probably needs to happen